### PR TITLE
#10860: Split Conv2dConfig into Compute config

### PR DESCRIPTION
**Pull request by sankarmanoj-tt**
### Ticket
#10860 

### Problem description
Conv2dConfig contains Compute kernel specific arguments which should be passed using DeviceComputeKernelConfig

### What's changed
Removed compute kernel arguments from Conv2dConfig. `conv2d` takes an additional argument called compute_config. Used pybind to expose helper functions to create DeviceComputeKernelConfig from the model code.

### Checklist
- [x] Post commit CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/11893281772)
- [x] Model regression CI testing [passes](https://github.com/tenstorrent/tt-metal/actions/runs/11948195753)
- [x] Demo tests [passes](https://github.com/tenstorrent/tt-metal/actions/runs/11935742809)
- [x] Device performance regression CI testing [passes](https://github.com/tenstorrent/tt-metal/actions/runs/11936504399)
- [x] Nightly CI run. [No new regressions. ](https://github.com/tenstorrent/tt-metal/actions/runs/12005424271)

### CODE REVIEW

File: ttnn/ttnn/device.py [hunk 1]
```diff
@@ -138,11 +138,37 @@ def is_grayskull(device):
...
+def GetComputeKernelConfig(
+    math_fidelity=ttnn.MathFidelity.LoFi,
+    math_approx_mode=True,
+    fp32_dest_acc_en=False,
+    packer_l1_acc=False,
+    dst_full_sync_en=False,
+):
+    device = ttnn.GetDefaultDevice()
+    if is_wormhole_b0(device) or is_blackhole(device):
+        return ttnn.WormholeComputeKernelConfig(
```
Comment (pavlejosipovic) [inline ttnn/ttnn/device.py:161]: I would create blackhole config here if it's blackhoel arch.
I know that currently they are just aliased, but it doesn't mean it will stay that way.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 2]
```diff
@@ -102,8 +102,10 @@ Result conv_transpose2d(
     std::array<uint32_t, 2> dilation,
     uint32_t groups,
     std::optional<const ttnn::Tensor> bias_tensor,
-    std::optional<const conv2d::Conv2dConfig> conv_config_)   {
+    std::optional<const conv2d::Conv2dConfig> conv_config_,
+    std::optional<const DeviceComputeKernelConfig> compute_config_)   {
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:106]: I would do pass by const ref for compute_config_, not just here but in general

File: ttnn/ttnn/device.py (hunk 1 above)
Comment (sankarmanoj-tt) [reply ttnn/ttnn/device.py:161]: Done.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 2 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:106]: I can make this change in a different PR.

Comment (pavlejosipovic) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:106]: Yeah I think there is PR in flight that does this in general 
https://github.com/tenstorrent/tt-metal/pull/15187

**Review by pavlejosipovic — COMMENTED** [review]
You need go over all instances of ttnn.Conv2dConfig() in the code and fixup params.
I see that many of them still accept math_fidelity for example

File: ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp [hunk 3]
```diff
@@ -58,7 +58,7 @@ Tensor optimized_conv_new(const Tensor& a, const Tensor &b, std::optional<const
     DataType dtype,
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:58]: there is no point in passing in here both math_fidelity and compute_kernel_config

Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp:59]: Removed math fidelity.

File: ttnn/ttnn/device.py [hunk 4]
```diff
@@ -138,11 +138,45 @@ def is_grayskull(device):
...
+        )
+    elif is_blackhole(device):
+        return ttnn.BlackholeComputeKernelConfig(
+            math_fidelity=math_fidelity,
+            fp32_dest_acc_en=fp32_dest_acc_en,
+            packer_l1_acc=packer_l1_acc,
+            math_approx_mode=math_approx_mode,
+            dst_full_sync_en=dst_full_sync_en,
+        )
+    else:
```
Comment (mywoodstock) [inline ttnn/ttnn/device.py:176]: I think we should check if `is_grayskull` instead of defaulting to it.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 5]
```diff
@@ -210,6 +184,19 @@ Result conv_transpose2d(
             0,
             input_tensor_post_tm.memory_config());
 
+        if(conv_config.deallocate_activation) {
+            input_tensor_post_tm.deallocate();
+            log_debug(tt::LogOp, "Deallocate Input Tensor");
+        }
+        if (conv_config.reallocate_halo_output) {
+            auto move_output = ttnn::operations::core::reallocate(halo_output, halo_output.memory_config());
+            ttnn::operations::core::deallocate(halo_output);
```
[the remark this answers is not in the document]
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:193]: I'm not certain. From a quick read through of the reallocate implementation, it looks like it does. I have removed the deallocate.

File: ttnn/ttnn/device.py (hunk 4 above)
Comment (sankarmanoj-tt) [reply ttnn/ttnn/device.py:176]: Done.

**Review by mywoodstock — APPROVED** [review]

**Comment by tt-rkim** [comment]
Please run nightly fast dispatch and perf pipelines and post the links in your PR

**Review by bbradelTT — APPROVED** [review]

**Review by esmalTT — APPROVED** [review]
Assuming model pipelines are passing - LGTM 👍

File: tests/ttnn/unit_tests/operations/test_small_resnet50_block.py [hunk 6]
```diff
@@ -118,6 +118,8 @@ def __call__(self, x, device, batch_size, input_height, input_width, conv_op_cac
             conv_config=ttnn.Conv2dConfig(
                 dtype=self.model_config["ACTIVATIONS_DTYPE"],
                 weights_dtype=self.model_config["WEIGHTS_DTYPE"],
+            ),
+            compute_config=ttnn.GetComputeKernelConfig(
```
Comment (pavlejosipovic) [inline tests/ttnn/unit_tests/operations/test_small_resnet50_block.py:122]: This will change math fidelity compared to the original code as defaults are not the same

**Review by pavlejosipovic — APPROVED** [review]

**Comment by tt-rkim** [comment]
No nightly

**Comment by sankarmanoj-tt** [comment]
@tt-rkim added Nightly CI run

File: conftest.py [hunk 7]
```diff
@@ -210,6 +210,8 @@ def mesh_device(request, silicon_arch_name, silicon_arch_wormhole_b0, device_par
     mesh_device = ttnn.open_mesh_device(mesh_shape, dispatch_core_type=get_dispatch_core_type(), **device_params)
 
     logger.debug(f"multidevice with {mesh_device.get_num_devices()} devices is created")
+    ttnn.SetDefaultDevice(mesh_device.get_device(0))
```
Comment (ayerofieiev-tt) [inline conftest.py:213]: @cfjchu , looks ok to you?

Comment (ayerofieiev-tt) [inline conftest.py:213]: Other functions around this one do not set the device as the default one. This change makes it inconsistent.

File: ttnn/ttnn/device.py [hunk 8]
```diff
@@ -138,11 +138,47 @@ def is_grayskull(device):
...
 
 
+def GetComputeKernelConfig(
+    math_fidelity=ttnn.MathFidelity.HiFi4,
+    math_approx_mode=True,
+    fp32_dest_acc_en=False,
+    packer_l1_acc=False,
+    dst_full_sync_en=False,
+):
+    device = ttnn.GetDefaultDevice()
```
Comment (ayerofieiev-tt) [inline ttnn/ttnn/device.py:159]: It is not guaranteed that a Default device is set. 
`GetComputeKernelConfig` seems to be used everywhere, making so much code rely on something that is not guaranteed.

Comment (ayerofieiev-tt) [reply ttnn/ttnn/device.py:159]: should this become a device's function?

should it accept device as a parameter? if not provided, function can try to query the default device and if it is not set - throw.

File: ttnn/ttnn/device.py [hunk 9]
```diff
@@ -138,11 +138,47 @@ def is_grayskull(device):
...
+
+
 SetDefaultDevice = ttnn._ttnn.device.SetDefaultDevice
 GetDefaultDevice = ttnn._ttnn.device.GetDefaultDevice
 format_input_tensor = ttnn._ttnn.device.format_input_tensor
 format_output_tensor = ttnn._ttnn.device.format_output_tensor
 pad_to_tile_shape = ttnn._ttnn.device.pad_to_tile_shape
 
 
+def GetComputeKernelConfig(
```
Comment (ayerofieiev-tt) [inline ttnn/ttnn/device.py:152]: This function does not get a config, it creates it. Should it be really called "get"?

File: ttnn/ttnn/device.py [hunk 10]
```diff
@@ -138,11 +138,47 @@ def is_grayskull(device):
...
+        )
+    elif is_blackhole(device):
+        return ttnn.BlackholeComputeKernelConfig(
+            math_fidelity=math_fidelity,
+            fp32_dest_acc_en=fp32_dest_acc_en,
+            packer_l1_acc=packer_l1_acc,
+            math_approx_mode=math_approx_mode,
+            dst_full_sync_en=dst_full_sync_en,
+        )
+    elif is_grayskull(device):
```
Comment (ayerofieiev-tt) [inline ttnn/ttnn/device.py:176]: instead of a series of if-else this could be a switch statement over device arch

File: conftest.py (hunk 7 above)
Comment (sankarmanoj-tt) [reply conftest.py:213]: An alternate approach: https://github.com/tenstorrent/tt-metal/commit/79031d8dd99bcb67116c4b89ba1d5f5cb96f1d25

Comment (sankarmanoj-tt) [reply conftest.py:213]: This checks the default device first, but if that's not set, checks the `ARCH_NAME` env flag.

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp [hunk 11]
```diff
@@ -211,6 +185,18 @@ Result conv_transpose2d(
...
+            log_debug(tt::LogOp, "Deallocate Input Tensor");
+        }
+        if (conv_config.reallocate_halo_output) {
+            auto move_output = ttnn::operations::core::reallocate(halo_output, halo_output.memory_config());
+            halo_output = move_output;
+            log_debug(tt::LogOp, "Reallocate Halo Output");
+        }
+
+
+
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp:199]: no need for excessive breaks

File: models/demos/convnet_mnist/tt/convnet_mnist.py [hunk 12]
```diff
@@ -19,19 +19,20 @@ def convnet_mnist(
...
-        fp32_dest_acc_enabled=False,
-        packer_l1_accum_enabled=False,
         input_channels_alignment=32,
         transpose_shards=False,
         reshard_if_not_optimal=True,
         deallocate_activation=True,
         reallocate_halo_output=True,
     )
-
+    compute_config = ttnn.CreateComputeKernelConfig(
```
Comment (ayerofieiev-tt) [inline models/demos/convnet_mnist/tt/convnet_mnist.py:30]: `def convnet_mnist(` receives device as parameter, but relies of some default set device to get compute config. This is not good.

**Review by ayerofieiev-tt — COMMENTED** [review]
I believe solving this on Python side is not the right place.
We need this same logic on C++ side too, right?
And we can't rely on globally set device, which is still what's happening here.

File: tt_metal/impl/buffers/circular_buffer_types.cpp [hunk 13]
```diff
@@ -99,13 +99,13 @@ CircularBufferConfig& CircularBufferConfig::set_globally_allocated_address_and_t
...
-            "size of {} B",
-            total_size,
-            this->max_size_);
+    // if (total_size > this->max_size_) {
+    //     TT_ASSERT(
+    //         false,
+    //         "Cannot set to globally allocated buffer. Circular buffer size {} B exceeds allocated L1 buffer bank "
+    //         "size of {} B",
+    //         total_size,
+    //         this->max_size_);
```
Comment (tt-aho) [inline tt_metal/impl/buffers/circular_buffer_types.cpp:108]: Revert this change

**Review by tt-rkim — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp (hunk 11 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp]: Removed.

File: tt_metal/impl/buffers/circular_buffer_types.cpp (hunk 13 above)
Comment (sankarmanoj-tt) [reply tt_metal/impl/buffers/circular_buffer_types.cpp]: Reverted.

**Review by ayerofieiev-tt — APPROVED** [review]
Please after this PR look at `init_device_compute_kernel_config` and see how to remove it.

I assume we just have to construct/build the DeviceKernelConfig. Having method like "init_device.." in ttnn is very misleading. It almost looks like I must call it once to initialize something and thats it.

**Review by shwetankTT — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]

**Review by uaydonat — APPROVED** [review]
